src/Slacklinker/Sender/UserDataUpload.hs

```diff
@@ -0,0 +1,108 @@
+module Slacklinker.Sender.UserDataUpload (doUploadUserData) where
+
+import Control.Monad.Catch (MonadThrow (..))
+import Data.Aeson (eitherDecode)
+import Database.Persist
+import Network.URI (parseURI)
+import OpenTelemetry.Instrumentation.HttpClient (
+  Request (..),
+  httpClientInstrumentationConfig,
+  httpLbs,
+  requestFromURI,
+  responseBody,
+  setRequestCheckStatus,
+ )
+import Slacklinker.App (App (..), HasApp (..), appSlackConfig, runDB)
+import Slacklinker.Import
+import Slacklinker.Models
+import Slacklinker.Sender.Internal
+import Slacklinker.Sender.Types
+import Slacklinker.Sender.UserDataUpload.Types
+import Slacklinker.Settings (SlacklinkerSetting (..), SlacklinkerSettingTag (..), getSetting)
+import Slacklinker.Types (SlackToken (..))
+import Web.Slack (usersListAll)
+import Web.Slack.Conversation
+import Web.Slack.Files.Types (FileMode (..), FileObject (..), FileObjectVisible (..))
+import Web.Slack.Pager (loadingPage)
+import Web.Slack.User
+import Web.Slack.User qualified as User
+
+updateSlackUsersList :: (HasApp m, MonadIO m) => WorkspaceMeta -> m ()
+updateSlackUsersList wsInfo = do
+  slackConfig <- appSlackConfig wsInfo.token
+  users_ <- liftIO $ usersListAll slackConfig (def {User.listReqLimit = Just 100})
+  users <- liftIO $ loadingPage users_ convertPage
+
+  runDB $ for_ users \user ->
+    let email = user.userProfile >>= (.profileEmail)
+     in upsertBy
+          (UniqueKnownUser wsInfo.workspaceId user.userId)
+          KnownUser
+            { workspaceId = wsInfo.workspaceId
+            , slackUserId = user.userId
+            , email
+            , emoji = Nothing
+            , githubUsername = Nothing
+            }
+          [KnownUserEmail =. email]
+
+  pure ()
+  where
+    convertPage resp =
+      fromList @(Vector _) <$> fromEither resp
+
+updateUserData :: MonadIO m => WorkspaceId -> UserDataUpload -> SqlPersistT m ()
+updateUserData wsId UserDataUpload {items} = do
+  for_ items \item -> do
+    updateWhere
+      [KnownUserWorkspaceId ==. wsId, KnownUserEmail ==. Just item.email]
+      [KnownUserEmoji =. item.callsign, KnownUserGithubUsername =. item.gitHubUsername]
+
+doUploadUserData :: (MonadThrow m, HasApp m, MonadUnliftIO m) => WorkspaceMeta -> ConversationId -> FileObject -> m ()
+doUploadUserData wsInfo cid file = do
+  ~(SettingAllowUploadUserData permitted) <- runDB $ getSetting @'AllowUploadUserData
```

Comment:
Just curious, what's the value of the GADT settings thing over just using an `allowUploadUserData` field in a record?

Reply:
The main reason is that it generates a `select` for that specific one at the sql level, so it's a nicer model of the actual SQL. I'm not super happy with how annoying it is to add new ones though. Feels like I may have overengineered it a bit.